java-squid/src/test/java/org/sonar/java/se/symbolicvalues/SymbolicValueRelationsTest.java

```diff
@@ -0,0 +1,619 @@
+      assertThat(relation.rightOp).isEqualTo(symmetricOfSymmetric.rightOp);
+    } catch (Exception e) {
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  private void checkImplies(Class<? extends SymbolicValueRelation> hypothesisClass, Class<? extends SymbolicValueRelation> checkedClass, RelationState directResult,
+    RelationState transposedResult) {
+    SymbolicValue a = values[0];
+    SymbolicValue b = values[1];
+    try {
+      SymbolicValueRelation hypothesis = createRelation(hypothesisClass, a, b);
+      SymbolicValueRelation checked = createRelation(checkedClass, a, b);
+      checkImplies(hypothesis, checked, directResult);
+      checked = createRelation(checkedClass, b, a);
+      checkImplies(hypothesis, checked, transposedResult);
+    } catch (Exception e) {
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  private void checkImplies(SymbolicValueRelation hypothesis, SymbolicValueRelation checked, RelationState expected) {
+    RelationState actual = hypothesis.implies(checked);
+    assertThat(actual).as(hypothesis.toString() + " => " + checked.toString()).isEqualTo(expected);
+  }
+
+  private void checkTransitive(Class<? extends SymbolicValueRelation> hClass1, Class<? extends SymbolicValueRelation> hClass2, Class<? extends SymbolicValueRelation> checkedClass,
+    RelationState directResult) {
+    SymbolicValue a = values[0];
+    SymbolicValue b = values[1];
+    SymbolicValue c = values[2];
+    try {
+      SymbolicValueRelation hypothesis1 = createRelation(hClass1, a, b);
+      SymbolicValueRelation hypothesis2 = createRelation(hClass2, b, c);
+      SymbolicValueRelation checked = createRelation(checkedClass, a, c);
+      checkTransitive(hypothesis1, hypothesis2, checked, directResult);
+      checkTransitive(hypothesis1.symmetric(), hypothesis2, checked, directResult);
+      checkTransitive(hypothesis1, hypothesis2.symmetric(), checked, directResult);
+      checkTransitive(hypothesis1, hypothesis2, checked.symmetric(), directResult);
+    } catch (Exception e) {
+      Assert.fail(e.getMessage());
+    }
+  }
+
+  private SymbolicValueRelation createRelation(Class<? extends SymbolicValueRelation> clazz, SymbolicValue a, SymbolicValue b) throws Exception {
+    Constructor<? extends SymbolicValueRelation> constructor = clazz.getDeclaredConstructor(SymbolicValue.class, SymbolicValue.class);
+    constructor.setAccessible(true);
+    return constructor.newInstance(a, b);
+  }
+
+  private void checkTransitive(SymbolicValueRelation hypothesis1, SymbolicValueRelation hypothesis2, SymbolicValueRelation checked, RelationState expected) {
+    List<SymbolicValueRelation> hypotheses = ImmutableList.of(hypothesis1, hypothesis2);
+    RelationState actual = checked.resolveState(hypotheses);
+    assertThat(actual).as(hypothesis1.toString() + " && " + hypothesis2.toString() + " => " + checked.toString()).isEqualTo(expected);
+  }
+
+  @Test
+  public void testEqualImplies() {
```

Comment:
you might want to have a look at parametrized test for all these checkImplies (which IMO should be named assertImplies) : parametrized tests allow you to run a matrix of data  through the same tests, see for instance http://www.tutorialspoint.com/junit/junit_parameterized_test.htm